app/src/main/kotlin/co/netguru/baby/monitor/client/feature/voiceAnalysis/RecordingController.kt

```diff
@@ -0,0 +1,63 @@
+package co.netguru.baby.monitor.client.feature.voiceAnalysis
+
+import co.netguru.baby.monitor.client.feature.machinelearning.MachineLearning
+import co.netguru.baby.monitor.client.feature.noisedetection.NoiseDetector
+import io.reactivex.Observable
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+import javax.inject.Inject
+
+class RecordingController @Inject constructor(
+    private val aacRecorder: AacRecorder
+) {
+    var voiceAnalysisOption: VoiceAnalysisOption = VoiceAnalysisOption.MACHINE_LEARNING
+
+    fun startRecording(): Observable<RecordingData> {
+        return aacRecorder.startRecording()
+            .scan(emptyArray<Byte>() to emptyArray<Short>(), { acc, new ->
+                if (acc.second.size >= MachineLearning.DATA_SIZE) {
+                    clearData()
+                } else {
+                    addData(acc, new)
+                }
+            })
+            .filter { it.first.isNotEmpty() && it.second.isNotEmpty() }
+            .map {
+                val data = when {
+                    isMachineLearningWithEnoughData(it) -> RecordingData.MachineLearning(
+                        it.first.toByteArray(),
+                        it.second.toShortArray()
+                    )
+                    isNoiseDetectionWithEnoughData(it) -> RecordingData.NoiseDetection(
+                        it.second.takeLast(NoiseDetector.DATA_SIZE).toShortArray()
+                    )
+                    else -> RecordingData.Raw(it.first.toByteArray())
+                }
+                data
+            }
+    }
+
+    private fun isNoiseDetectionWithEnoughData(it: Pair<Array<Byte>, Array<Short>>) =
+        voiceAnalysisOption == VoiceAnalysisOption.NOISE_DETECTION &&
+                it.second.size >= NoiseDetector.DATA_SIZE
+
+    private fun isMachineLearningWithEnoughData(it: Pair<Array<Byte>, Array<Short>>) =
+        voiceAnalysisOption == VoiceAnalysisOption.MACHINE_LEARNING &&
+                it.second.size == MachineLearning.DATA_SIZE
+
+    private fun addData(
+        accumulator: Pair<Array<Byte>, Array<Short>>,
+        newData: ByteArray
+    ): Pair<Array<Byte>, Array<Short>> {
+        return accumulator.first.plus(newData.toTypedArray()) to accumulator.second
+            .plus(bytesToShorts(newData).toTypedArray())
+    }
+
+    private fun clearData() = Pair(emptyArray<Byte>(), emptyArray<Short>())
+
+    private fun bytesToShorts(bytes: ByteArray): ShortArray {
+        val shorts = ShortArray(bytes.size / 2)
```

Comment:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/3e96add8605b778f23b492b61de8a8adbd52836c/app/src/main/kotlin/co/netguru/baby/monitor/client/feature/voiceAnalysis/RecordingController.kt#L59"></span>This expression contains a magic number. Consider defining it to a well named constant.</td>
    </tr>
  </tbody>
</table>